Talk image attachments crash Viewer after first open: undefined file id/mime in FilePreview.vue

<!--- Please keep this note for other contributors -->
> [!TIP]
> ### How to use GitHub
>
> * Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
> * Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
> * Subscribe to receive notifications on status change and new comments.

---
<!-- DO NOT EDIT ABOVE THIS LINE -->

## Steps to reproduce

1. Open Talk in the Nextcloud web UI.
2. Upload an image into a chat.
3. Click the image preview in the chat.
4. The image opens in the viewer.
5. Close the viewer.
6. Click the same image preview again.

### Expected behaviour

Talk image attachments should be openable repeatedly in the image viewer/fullscreen view, just like images opened from the regular Nextcloud Files app.

### Actual behaviour

Image attachments shown in a Talk chat can often only be opened once, or sometimes not at all.

The thumbnail/preview is visible and the file can still be downloaded, but opening it in the image viewer/fullscreen view fails after the first attempt.

Sometimes a browser reload allows the image to be opened once again, but then the same failure repeats.

The same image opens normally and repeatedly from the regular Nextcloud Files app:

```text
Files -> Talk folder -> same image -> open -> close -> open again
```

This works repeatedly without the Talk viewer crash.

Therefore the issue seems specific to Talk's file preview component or the way Talk passes file data to the Viewer app.

## Talk app

**Talk app version:** `23.0.4`

**Custom Signaling server configured:** yes, Nextcloud AIO Talk/signaling setup

**Custom TURN server configured:** yes, Nextcloud AIO Talk/TURN setup

**Custom STUN server configured:** yes, Nextcloud AIO Talk/STUN setup

## Browser

**Microphone available:** yes

**Camera available:** yes

**Operating system:** Linux

**Browser name:** Chrome

**Browser version:** `148.0.0.0`

### Browser log

<details>

```text
browserCheck.ts:74 Detected browser Chrome 148 (148.0.0.0)

Viewer.js:89 [ERROR] viewer: Could not register handler
{
  app: "viewer",
  uid: "<USER_ID>",
  level: 2,
  error: "The handler is already registered",
  handler: {...}
}

GET https://cloud.example.tld/ocs/v2.php/profile/<GUEST_ID> 400 (Bad Request)

AxiosError: Request failed with status code 400
    at getUserProfileInformation (groupware.ts:179:22)
    at RightSidebarContent.vue:197:2

signaling.js:1177 Not connected to signaling server yet, defer joining room <ROOM_TOKEN>

TypeError: Cannot read properties of undefined (reading 'mime')
    at o.openFileFromList (Viewer.vue:816:1)
    at o.next (Viewer.vue:1078:1)
    at o.next (NcModal.mjs:484:14)

Viewer.vue:481 [WARN] viewer: No files provided, skipping update
{
  app: "viewer",
  uid: "<USER_ID>",
  level: 2
}

useViewer.js:89 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'id')
    at n (useViewer.js:89:27)
    at Array.map (<anonymous>)
    at Proxy.a (useViewer.js:114:15)
    at Proxy.handleClick (FilePreview.vue:527:10)
```

The most relevant part seems to be:

```text
FilePreview.vue:527
useViewer.js:89
Cannot read properties of undefined (reading 'id')
Cannot read properties of undefined (reading 'mime')
No files provided, skipping update
```

This looks like Talk's `FilePreview.vue` passes an undefined or incomplete file object/list to the Viewer.

</details>

## Server configuration
<!--
You can use the Issue Template application to prefill most of the required information: https://apps.nextcloud.com/apps/issuetemplate
-->

**Operating system**: Debian / Docker host

**Web server:** Nextcloud AIO Apache container behind reverse proxy

**Database:** PostgreSQL

**PHP version:** Nextcloud AIO default PHP version

**Nextcloud Version:** `33.0.3.2`

**List of activated apps:**

<details>

```text
- files: 2.5.0
- files_sharing: 1.25.2
- files_versions: 1.26.0
- files_pdfviewer: 6.0.0
- spreed: 23.0.4
- viewer: 6.0.0
```

Additional relevant checks:

```bash
php occ integrity:check-app spreed
php occ integrity:check-app viewer
```

Both commands returned no output/errors.

</details>

**Nextcloud configuration:**

<details>

```text
Deployment: Nextcloud AIO behind reverse proxy
Public URL sanitized as: https://cloud.example.tld
```

The issue is reproducible in Talk, but the same file opens normally from the regular Files app.

</details>

### Server log (data/nextcloud.log)

<details>

```text
There was initially a server-side preview error for one old file:

GET /core/preview?fileId=<FILE_ID>&x=-1&y=384&a=1

SQLSTATE[23505]: Unique violation:
duplicate key value violates unique constraint "previews_file_uniq_idx"

Key (file_id, width, height, mimetype_id, cropped, version_id)
=(<FILE_ID>, 256, 973, 9, f, -1) already exists.
```

This was investigated and cleaned up:

```bash
php occ config:system:set preview_concurrency_new --type=integer --value=1
php occ config:system:set preview_concurrency_all --type=integer --value=2

rm -rf /mnt/ncdata/appdata_<INSTANCE_ID>/preview
delete from oc_previews;
php occ files:scan-app-data
```

After cleanup, the affected old file no longer existed in `oc_previews`:

```sql
select id,file_id,width,height,mimetype_id,cropped,version_id
from oc_previews
where file_id=<FILE_ID>;
```

Result:

```text
(0 rows)
```

The Talk viewer issue still persisted afterwards.

Frontend asset cache was also cleared:

```bash
rm -rf /mnt/ncdata/appdata_<INSTANCE_ID>/js
rm -rf /mnt/ncdata/appdata_<INSTANCE_ID>/css
```

Browser was hard-reloaded afterwards. The issue persisted.

The `files_downloadlimit` app was initially enabled with:

```json
{
  "enabled": "yes",
  "default-download-limit": 2
}
```

It was disabled for testing:

```bash
php occ app:disable files_downloadlimit
```

The issue still occurred, so this does not seem to be caused by download limits.

Attempting to disable Viewer:

```bash
php occ app:disable viewer
```

Result:

```text
viewer can't be disabled.
```

</details>

## Additional notes

- The same file opens normally from the Files app.
- The failure occurs when clicking the image preview inside Talk.
- The console stack points to:
  - `FilePreview.vue`
  - `useViewer.js`
  - `Viewer.vue`
- Errors indicate that `undefined` is passed where file info/list entries are expected:
  - missing `id`
  - missing `mime`
  - "No files provided, skipping update"

This seems similar to known Viewer issues where the viewer receives an incomplete file list or file object, but in this case the trigger is Talk's file preview component.

Could you confirm whether this is a known regression in Talk `23.0.4` / Nextcloud `33.0.3`?

--- a/src/stores/__tests__/sharedItems.spec.js
+++ b/src/stores/__tests__/sharedItems.spec.js
@@ -23,6 +23,15 @@ vi.mock('vuex', () => ({
 	})),
 }))
 
+function requiresObject(type) {
+	return [
+		SHARED_ITEM.TYPES.LOCATION,
+		SHARED_ITEM.TYPES.DECK_CARD,
+		SHARED_ITEM.TYPES.POLL,
+		SHARED_ITEM.TYPES.OTHER,
+	].includes(type)
+}
+
 describe('sharedItemsStore', () => {
 	const token = 'TOKEN'
 	let sharedItemsStore
@@ -35,8 +44,10 @@ describe('sharedItemsStore', () => {
 		setActivePinia(createPinia())
 		sharedItemsStore = useSharedItemsStore()
 		sharedItemsOrder.forEach((type, index) => {
-			payloadOverview[type] = [{ id: 100 + index, message: type }]
-			result[type] = { [100 + index]: { id: 100 + index, message: type } }
+			const message = { id: 100 + index, message: type, messageParameters: {} }
+			message.messageParameters = requiresObject(type) ? { object: {} } : { file: {} }
+			payloadOverview[type] = [message]
+			result[type] = { [100 + index]: message }
 		})
 	})
 

--- a/src/stores/sharedItems.ts
+++ b/src/stores/sharedItems.ts
@@ -20,9 +20,9 @@ import {
 	unpinMessage,
 } from '../services/messagesService.ts'
 import { getSharedItems, getSharedItemsOverview } from '../services/sharedItemsService.ts'
-import { getItemTypeFromMessage } from '../utils/getItemTypeFromMessage.ts'
+import { getItemTypeFromMessage, isValidSharedItem } from '../utils/getItemTypeFromMessage.ts'
 
-type SharedItemType = keyof SharedItemsOverview
+type SharedItemType = typeof SHARED_ITEM.TYPES[keyof typeof SHARED_ITEM.TYPES] | (string & {})
 
 type SharedItemsPoolType = Record<string, {
 	[K: string]: Record<number, SharedItems[keyof SharedItems]>
@@ -78,7 +78,9 @@ export const useSharedItemsStore = defineStore('sharedItems', () => {
 			if (Object.keys(data[type]).length) {
 				checkForExistence(token, type)
 				for (const message of data[type]) {
-					sharedItemsPool[token][type][message.id] = message
+					if (isValidSharedItem(type, message)) {
+						sharedItemsPool[token][type][message.id] = message
+					}
 				}
 			}
 		}
@@ -96,7 +98,9 @@ export const useSharedItemsStore = defineStore('sharedItems', () => {
 	function addSharedItemFromMessage(token: string, message: ChatMessage, type?: SharedItemType) {
 		const itemType = type ?? getItemTypeFromMessage(message)
 		checkForExistence(token, itemType)
-		sharedItemsPool[token][itemType][message.id] = message
+		if (isValidSharedItem(itemType, message)) {
+			sharedItemsPool[token][itemType][message.id] = message
+		}
 	}
 
 	/**
@@ -135,7 +139,9 @@ export const useSharedItemsStore = defineStore('sharedItems', () => {
 		checkForExistence(token, type)
 
 		messages.forEach((message) => {
-			sharedItemsPool[token][type][message.id] = message
+			if (isValidSharedItem(type, message)) {
+				sharedItemsPool[token][type][message.id] = message
+			}
 		})
 	}
 

--- a/src/utils/getItemTypeFromMessage.ts
+++ b/src/utils/getItemTypeFromMessage.ts
@@ -1,9 +1,10 @@
-import type { ChatMessage } from '../types/index.ts'
-
 /**
  * SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
  * SPDX-License-Identifier: AGPL-3.0-or-later
  */
+
+import type { ChatMessage, PinnedChatMessage } from '../types/index.ts'
+
 import { MESSAGE, SHARED_ITEM } from '../constants.ts'
 
 /**
@@ -38,4 +39,43 @@ export function getItemTypeFromMessage(message: ChatMessage): string {
 	} else {
 		return SHARED_ITEM.TYPES.OTHER
 	}
+}
+
+/**
+ * Validates whether a shared item has the required messageParameters for its type.
+ * Only valid items should be stored and rendered.
+ *
+ * @param type shared item type
+ * @param message message to validate
+ * @return true if the message has valid messageParameters for its type
+ */
+export function isValidSharedItem(type: string, message: ChatMessage | PinnedChatMessage): boolean {
+	// Pinned messages have a different structure
+	if (type === SHARED_ITEM.TYPES.PINNED) {
+		return true
+	}
+
+	// Items that require messageParameters.object
+	if ([
+		SHARED_ITEM.TYPES.LOCATION,
+		SHARED_ITEM.TYPES.DECK_CARD,
+		SHARED_ITEM.TYPES.POLL,
+		SHARED_ITEM.TYPES.OTHER,
+	].includes(type)) {
+		return !!(message.messageParameters?.object)
+	}
+
+	// Items that require messageParameters.file
+	if ([
+		SHARED_ITEM.TYPES.FILE,
+		SHARED_ITEM.TYPES.AUDIO,
+		SHARED_ITEM.TYPES.MEDIA,
+		SHARED_ITEM.TYPES.RECORDING,
+		SHARED_ITEM.TYPES.VOICE,
+	].includes(type)) {
+		return !!(message.messageParameters?.file)
+	}
+
+	// At least one truthy property should be present
+	return !!(message.messageParameters?.object) || !!(message.messageParameters?.file)
 }

